apps/aecore/lib/aecore/chain/chain_state.ex

```diff
@@ -1,50 +1,50 @@
 defmodule Aecore.Chain.ChainState do
   @moduledoc """
-  Module used for calculating the block and chain states.
+  Module used for calculating the chain states.
   The chain state is a map, telling us what amount of tokens each account has.
   """
 
   require Logger
+  alias Aecore.Structures.SignedTx
 
-  @doc """
-  Calculates the balance of each account mentioned
-  in the transactions a single block, returns a map with the
-  accounts as key and their balance as value.
-  """
-  @spec calculate_block_state(list(), integer()) :: map()
-  def calculate_block_state(txs, latest_block_height) do
-    block_state = %{}
-
-    block_state =
-      for transaction <- txs do
-        updated_block_state =
-          cond do
-            transaction.data.from_acc != nil ->
-              update_block_state(block_state, transaction.data.from_acc,
-                                 -(transaction.data.value + transaction.data.fee),
-                                 transaction.data.nonce, transaction.data.lock_time_block, false)
-
-            true ->
-              block_state
-          end
-
-        add_to_locked = latest_block_height + 1 <= transaction.data.lock_time_block
-
-        update_block_state(updated_block_state, transaction.data.to_acc, transaction.data.value,
-                           0, transaction.data.lock_time_block, add_to_locked)
-      end
-
-    reduce_map_list(block_state)
+  @spec calculate_and_validate_chain_state!(list(), map(), integer()) :: map()
+  def calculate_and_validate_chain_state!(txs, chain_state, block_height) do
+    Enum.reduce(txs, chain_state, fn(transaction, chain_state) ->
+      apply_transaction_on_state!(transaction, chain_state, block_height) 
```

Comment:
Shouldn't we try-catch here as well?

Reply:
no, any exception in `apply_transaction_on_state!` should be handled by function calling `calculate_and_validate_chain_state!`